---
title: "GS in CIMMYT Wheat"
author: "Make this your own"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---
#Objective

Set up a genomic selection for CIMMYT wheat trials.  The data included are a hapmap file that has previously been imputed using Beagle.  The phenotype data file consist of 3 site years of data, 14_OBR_EYT--2014 Drought, 15_OBR_EYT -- 2015 Drought, 14_OBR_EYTHeat -- 2015 heat.  Note that not all phenotypes were included each year.  The lines were grouped into trials, with trials composing sister (highly related lines).

The phenotype information includes:  
GID_DNA:  This is the individual unique number that links to the hapmap file
blup_gs1:  Model estimated value for NDVI 1st measurement 	
blup_gs2:	Model estimated value for NDVI 2nd measurement
blup_gs3: Model estimated value for NDVI 3rd measurement	
blup_n1:	Model estimated value for canopy temperature 1st measurement
blup_n2:	Model estimated value for canopy temperature 2nd measurement
blup_n3:	Model estimated value for canopy temperature 3rd measurement
Experiment_id:  Identifies year and type of experiment heat or drought	
Entry:  Entry number	
cimmyt_tid:  CIMMYT Trial ID, grouping of individuals into trial	
notes:  notes	
gy_effect:  Estimated BLUP (0 centered) of grain yield	
gs: average of blup_gs1, blup_gs2, blup_gs3, and blup_gs4	
ir: average of blup_n1, blup_n2, blup_n3, blup_n4	
model: information about the model used to calcuate trial BLUPs	
blup_gs4:	Model estimated value for NDVI 4th measurement
blup_n4: Model estimated value for canopy temperature 4th measurement

This data is from Crain et al. (2018), Combining High-Throughput Phenotyping and
Genomic Information to Increase Prediction and Selection Accuracy in Wheat Breeding.  The data sets and scripts are stored at https://datadryad.org/resource/doi:10.5061/dryad.7f138.

#Objective 1:  Set up a model using only the drought 2014 data for grain yield.

```{r setup, eval=TRUE}
knitr::opts_knit$set(root.dir = '/Users/jcrain/Documents/Class/Genomic_Selection/Github/BecA/Trial_Data/')
setwd('/Users/jcrain/Documents/Class/Genomic_Selection/Github/BecA/Trial_Data/')
options(digits = 10) #set options to include lots of digits
#Git link
#https://github.com/jcrain1/BecA
```

```{r get_functions}
####Function to count alleles and populations parameters###
tassel5_to_params=function(x="hap matrix", y="columns to skip", z="population number"){
      geno=x
      #recount allele A and B and het
      alleleA=rowSums(geno[,(y+1):ncol(geno)]!=substring(geno$alleles, 3, 3) & geno[,(y+1):ncol(geno)]!="N") #only counts what is not allele B and missing.  i.e. counts allele A and various calls for heterozygous
      alleleB=rowSums(geno[,(y+1):ncol(geno)]!=substring(geno$alleles, 1, 1) & geno[,(y+1):ncol(geno)]!="N")
      het=rowSums(geno[,(y+1):ncol(geno)] == "M") + rowSums( geno[,(y+1):ncol(geno)] ==   "R") + rowSums(geno[,(y+1):ncol(geno)] ==  "W") + rowSums(geno[,(y+1):ncol(geno)] ==  "K") + rowSums(geno[,(y+1):ncol(geno)] ==  "S") + rowSums(geno[,(y+1):ncol(geno)] ==  "Y")
      present=1-(rowSums(geno[,(y+1):ncol(geno)]=="N")/z)
      MAF=apply(cbind(((alleleA-het)*2+het), (alleleB-het)*2+het), 1, min)/apply(
    cbind(((alleleA-het)*2+het), ((alleleB-het)*2+het)), 1, sum) 
      percentHet=het/apply(cbind(alleleA-het, alleleB-het, het), 1, sum)
      return(cbind.data.frame(geno[,1:y], "alleleA"=alleleA, "alleleB"=alleleB, "het"=het, "present"= present, "MAF"=MAF, "percentHET"=percentHet, geno[,(y+1):ncol(geno)]))
}

##function to convert hap to 0 and 1
hap_to_G=function(x="hap matrix", y="number of columns of information"){
  ##From Prasana, pulls out first allele for a and second for b
  a = substring(x$alleles,1,1)
  #Checks the frequency of the alleles if the second allele is more frequent it is substitued
  a[x$alleleA<x$alleleB] = substring(x$alleles,3,3)[x$alleleA<x$alleleB]
  #Same thing with the second allele
  b = substring(x$alleles,3,3)
  b[x$alleleA<x$alleleB] = substring(x$alleles,1,1)[x$alleleA<x$alleleB]
  #Checks to make sure all alleles are one or the other
  #print(paste("If 0 all alleles are accounted for: ", sum(a == b), sep=""))
  
  ## Turn into letter matrix for mapping
  #makes a copy of the hap matrix
  hap01 = x
  #sets all allele values to NA
  hap01[,(y+1):ncol(hap01)]=NA
  
  ## Turn allele a and allele b into 1 and -1.  Het into 0
  #line by line if a line is a then it places 1 in hap01 for the allele
  hap01[x == a] = 1
  hap01[x == b] = -1
  hap01[x == "M"] = 0
  hap01[x == "Y"] = 0
  hap01[x == "K"] = 0
  hap01[x == "R"] = 0
  hap01[x == "W"] = 0
  hap01[x == "S"] = 0
  hap01[x== "N"]=NA
  
  return(hap01)}
```

#Load Data

```{r load_data}
marker <- read.delim(file = './data/Original_Data/Beagle_Imputed.hmp.txt', check.names = FALSE, header = TRUE, stringsAsFactors = FALSE) #load marker data

phen <- read.table(file = './data/Original_Data/EYT_Phenotype_Imputed_GS_Simple.txt', header = TRUE, stringsAsFactors = FALSE) #load phenoytpe data

```

#Examine Data
```{r examine_data}
#potential genotype filtering
hap <- tassel5_to_params(marker, 11, z = 2033) #get population paramters

hist(hap$MAF) #there are few alleles that could be filtered
hist(hap$present) #why is this?  Imputed data for all values
hist(hap$percentHET) #inbred wheat lines is this expected?

hap <- hap[hap$MAF > 0.05, ] #filter at MAF > 0.05

geno <- hap_to_G(hap, 17) #make marker matrix

rownames(geno) <- geno$`rs#` #give rownames
gs_marker <- t(geno[, 18:ncol(geno)])

dim(gs_marker) #check gs_marker number

#potential phenotype filtering
str(phen)
#only get 2014 drought
phen <- phen[phen$Experiment_id == '14_OBR_EYT', ]
phen <- phen[, c(1, 8:10, 12)] #only get grain yield, trial_id, GID etc.  Just easier to work with less mess
any(is.na(phen$gy_effect)) #check that all grain yield is ther

#get only marker in phen
gs_marker <- gs_marker[rownames(gs_marker) %in% phen$GID_DNA, ]

nrow(gs_marker) == nrow(phen) #check number of individuals in marker and phenotype population the same
```

#Set up GS Model
```{r gs_model}
require(rrBLUP)
gs_pred = NULL

for(i in 1:max(phen$cimmyt_tid)){ #enter for loop #MODIFIED 
    x = phen
    print(paste('Loop or CV fold:', i, sep=' ')) #check progress
      phenoPredict <- x[x$cimmyt_tid == i ,] #get phenoPredict group
      phenoTrain <- x[!x$cimmyt_tid == i , ] #get phenoTrain group
      
     #order marker matrix by rbind values of phenoTrain and phenoPredict
      ord <- c(as.character(phenoTrain$GID_DNA), as.character(phenoPredict$GID_DNA))

      #order genotypic data
      g <- gs_marker[match(ord, rownames(gs_marker)),]

      print(paste('Check that g has training individuals first: ', all(as.character(phenoTrain$GID_DNA) == rownames(g)[1:nrow(phenoTrain)] && !is.null(phenoTrain$GID_DNA)), sep='')) #check
      print(paste('Check g has prediction individuals last: ', all(as.character(phenoPredict$GID_DNA) == rownames(g)[(nrow(g)-nrow(phenoPredict)+1):nrow(g)]  && !is.null(phenoPredict$GID_DNA)) , sep=''))
      
        #rrBLUP for genomic additive relationship matrix
        mat <- A.mat(g)

     ##prepare phenotype frame
       phenoPredict[ , 'gy_effect'] <- NA #set phenotype value to NA

       ##make phenotypic frame
       phenotypic <- rbind(phenoTrain, phenoPredict)

    #check that phenotypic and genotypic order are correct
       print(paste('Check that all phenotype data is in order:  ', all(rownames(g) == phenotypic$GID_DNA && !is.null(phenotypic$GID_DNA)), sep=''))

        ans <- kin.blup(phenotypic, K = mat, geno = "GID_DNA" , pheno ="gy_effect") #run rrBLUP
     
      preds <- ans$pred #extract prediction

      preds <- preds[names(preds) %in% phenoPredict$GID_DNA] #get only predictions
      preds <- as.data.frame(preds) #make into a dataframe
      preds$GID_DNA <- rownames(preds) #get plant_id
      rownames(preds) <- NULL #set rownames as NULL
      
      #set out
      gs_pred <- rbind(gs_pred, preds)
      
} #close for loop




```

#Validate Model
```{r examine_model_accuracy}
#Evaluate prediction
accuracy <- merge(phen, gs_pred, by = 'GID_DNA') #merge prediction data set and original phenotyep data set

pdf(file = './data/Figures/GS_Accuracy_Plot.pdf', height = 6, width = 8)
plot(accuracy$gy_effect, accuracy$preds, xlab = 'Observed value', ylab = 'Predicted value') #quick graph
abline(lm(accuracy$preds~accuracy$gy_effect), col = 'red') #add trendline
dev.off()

cor(accuracy$gy_effect, accuracy$preds) #correlation or accuracy

```

#Objective 2 Make your own model for a trait and experiment

##GWAS on data
Implement a GWAS on data.

```{r gwas_on_data}
require(factoextra)
require(rrBLUP)

#run and save results
geno_pca <- prcomp(gs_marker, scale. = T) #run PCA

fviz_eig(geno_pca) #scree plot with factoextra

#extract components
std_dev <- geno_pca$sdev

#compute variance
pr_var <- std_dev^2
prop_varex <- pr_var/sum(pr_var) #compute proportion of the variance

plot(cumsum(prop_varex), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     type = "b")

screeplot(geno_pca, npcs = 25, type = 'line')

#make realized relationship matrix
gmat <- A.mat(gs_marker) #get realized relationship matrix

#organize phenotype data
all(phen$GID_DNA==rownames(gmat))
phen <- phen[match(rownames(gmat), phen$GID_DNA),]
all(phen$GID_DNA==rownames(gmat))

#organize markers to go into GWAS
gwas_marker <- t(gs_marker)
gwas_marker <- cbind(hap[, c(1, 3, 4)], gwas_marker) #join map information from hap file

#Run gwas model
rrout <- GWAS(phen[, c(1,5)], geno = gwas_marker, K = gmat, min.MAF = 0.05, n.core=3, plot = TRUE, n.PC = 9)


#Chromosome 5B 212342796bp -- 226781405bp  14438609 (14Mb region)
rrout[which.max(rrout$gy_effect), ] #get chromosome to investigate
```

#R Session Information
```{r what_packages_and_version}
sessionInfo() #get session information
```